File: 89.SegmentedSieveAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.util.*;
import java.io.*;

public class SegmentedSieve {

    // Step 1: Simple Sieve to find primes up to sqrt(R)
    static List<Integer> simpleSieve(int limit) {
        boolean[] sieve = new boolean[limit + 1];
        Arrays.fill(sieve, true);
        sieve[0] = sieve[1] = false;

        for (int i = 2; i * i <= limit; i++) {
            if (sieve[i]) {
                for (int j = i * i; j <= limit; j += i) {
                    sieve[j] = false;
                }
            }
        }

        List<Integer> primes = new ArrayList<>();
        for (int i = 2; i <= limit; i++) {
            if (sieve[i]) primes.add(i);
        }
        return primes;
    }

    // Step 2: Segmented Sieve to find primes in range [L, R]
    static List<Integer> segmentedSieve(long L, long R) {
        int limit = (int) Math.sqrt(R) + 1;
        List<Integer> primes = simpleSieve(limit);

        boolean[] isPrime = new boolean[(int) (R - L + 1)];
        Arrays.fill(isPrime, true);

        for (int p : primes) {
            long start = Math.max((long) p * p, ((L + p - 1) / p) * (long) p);
            for (long j = start; j <= R; j += p) {
                isPrime[(int) (j - L)] = false;
            }
        }

        if (L == 1) isPrime[0] = false;

        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < isPrime.length; i++) {
            if (isPrime[i]) result.add((int) (L + i));
        }

        return result;
    }

    public static void main(String[] args) {
        long L = 10;
        long R = 50;
        List<Integer> primesInRange = segmentedSieve(L, R);

        System.out.println("Prime numbers between " + L + " and " + R + ":");
        System.out.println(primesInRange);
    }
}

File: 89.SegmentedSieveAlgo.py
# The Segmented Sieve Algorithm is an optimized version of the Sieve of Eratosthenes 
# used to find all prime numbers in a given range [L, R], especially when R is very large. 
# It divides the range into smaller segments and uses precomputed small primes 
# to mark non-primes efficiently.

# Explanation:

# The Sieve of Eratosthenes efficiently finds all primes up to √R.
# The Segmented Sieve then uses these small primes to mark multiples in each segment [L, R].

# This approach avoids storing an unnecessarily large array when R is big (e.g., up to 10¹²).

# Steps:

# Find all primes up to √R using the normal sieve.

# Create a boolean array for the segment [L, R].

# For each small prime p, mark its multiples in [L, R] as non-prime.

# Remaining unmarked numbers in [L, R] are prime.

import math

def simple_sieve(limit):
    sieve = [True] * (limit + 1)
    sieve[0] = sieve[1] = False
    for num in range(2, int(math.sqrt(limit)) + 1):
        if sieve[num]:
            for multiple in range(num * num, limit + 1, num):
                sieve[multiple] = False

    return [i for i in range(2, limit + 1) if sieve[i]]

def segmented_sieve(L, R):
    limit = int(math.sqrt(R)) + 1
    primes = simple_sieve(limit)

    is_prime = [True] * (R - L + 1)

    for p in primes:
        start = max(p * p, (L + p - 1) // p * p)
        for j in range(start, R + 1, p):
            is_prime[j - L] = False

    if L == 1:
        is_prime[0] = False

    primes_in_range = [L + i for i in range(R - L + 1) if is_prime[i]]
    return primes_in_range

L = 10
R = 50
result = segmented_sieve(L, R)

print(f"Prime numbers between {L} and {R}:")
print(result)





